### DISCUSSION TITLE
Efficient way to generate large repetetive shapes

### DISCUSSION THREAD

**Discussion by kriho** [sgenoud/replicad#146] [Q&A] [answered]
Hi everyone,

what is the fastest (in terms of runtime) way to generate large repetetive shapes such as a honycomb pattern?
My findings so far:
1) fuse drawings first until one drawing is left, then sketch and extrude it
2) making a drawing for an entire row and then cloning the row for each column greately reduces operation time
3) caching drawings and shapes also helpes quite a lot - seems like cloning is a lot cheaper then generating the data

To stay in the honeycomb example, are there more efficient ways then using `drawPolysides()` MANY times while utilizing the findings listed above?

Thanks!

**Comment by raydeleu** [comment]
I tried to create a honeycomb pattern, see https://github.com/sgenoud/replicad/discussions/66. I did not succeed in creating a drawing of a honeycomb first and then extruding the complete drawing. So I created a large number of hexagonal column which I then subtracted from the original shape. Not very efficient but it works.

**Reply by kriho** [reply]
I did not have any trouble generating the pattern. 
This is an intricate example that also crates the smallest possible pattern that will fill a provided area: [workbench](https://studio.replicad.xyz/workbench#code=UEsDBAoAAAAIAJKaZVjvCcvRTgIAAGIHAAAHAAAAY29kZS5qc51VTY%252BbMBC98ytGK61kbwibDzWHpas9tJceqkZ7ao8sdhK0LqS2I0Cr%252FPfOYEj4aqKUA4yH55nxmzfgxVlqLHyA0FHuV%252FdXGdso3SrplutMlSYR0tRvs0MqpGiBvqI3SbdwhGfQcq%252BSOBKh5ylpIU%252BE3aF7MZuFlUPIfeWYNw6Ni5Uzc%252FIHC7fYqMji2uqDrGO9ZaJET6dCViXwYc4D8y5tvPuRrlWUSnb389cdD2SB%252B4Vkn3jo1dvpEWwORrJuoLnvarsWiLvytjoRGC2VeXN8hm%252BSDaO6OXx4gJejVhTECzzCAibwPbK7wCQpq4z1N3QvOTzg2Segw%252FYuKtah%252F2jLKpBGUN4GxUUDimWimGMDpqA5xsW8mJN38GUX77rxSMmmyOEAv8fW15SfVMC0Dyse6MxGVrLlrN5ApJhdpu2XTB1%252Bp0NuCLTJNCMgBZ3GZYjG52eojMmkYY2ubqTWyrWO6jph6QpilWGzeNdpdZQaRWXOfDrjA5RnRF3SsXVcjLEdTVpHH0k%252BSDySlCHtE6TXgXpUFBUVRYgGUUFGl4paafRw%252BZ10ojfDCg73qKqXduFP7cL5TSQVWKsofOgnIOlN8SSkjiF9R5DKyKHir2n3LPA4M%252F1x0M2odN%252BtRkelPwX1GNAE8LOur45BMzblDWPQ0%252F5rll8W%252FtVut8I05ll1l8QmqHvYvBn%252Fh7B7Qf9P0lUWhoSTninZmKSvTvdQ0mfJlbXkXk5FP52Kvk3Ow5gk4%252BqjeOFrcPS0tAedtn4UVGc3ycgvogto%252FhdL7MZfUEsBAhQACgAAAAgAkpplWO8Jy9FOAgAAYgcAAAcAAAAAAAAAAAAAAAAAAAAAAGNvZGUuanNQSwUGAAAAAAEAAQA1AAAAcwIAAAAA)

As you can see, the generation takes about 25 seconds. My question purely was about efficiency. Are there faster ways to do this?

**Reply by sgenoud** [reply, marked as answer]
I don't think the issue is the generation of the shapes themselves - but the boolean operations that tend to be slow (there are some optimisation that can be triggered, but they tend not to be that great).

Sometimes you cannot avoid them (in the case you show for instance). But, you can have a look at what I did with my implementation of the [honeycomb storage wall](https://github.com/sgenoud/replicad-hsw-generator/blob/main/src/cad/honeycomb.js) to see how you can build without booleans - and it is faster even with a more complex geometry.

**Reply by kriho** [reply]
I rely on the booleans as they are part of the user input. There is no way to eliminate them for me. 
Are there differences in efficency between shapes that are fused out of many shapes or a very complex single shape when it comes to booleans?
